### PR TITLE
feat: implementation of yearn router contract

### LINKED ISSUE #678: Point Yearn contract calls to Shapeshift DAO router contract instead of directly at the yearn vaults.
The Shapeshift router contract has been deployed. We now need to point at that instead of the yearn vaults.

### PR DESCRIPTION
**Pull request by theoboldfrazier**
## Description

Updates the implementation of yearn vault deposits and withdraws to go through the new Shapeshift router affiliate contract.

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?
- [x] Do all new and existing tests pass? Does the linter pass?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes #678 

## Testing

Please outline all testing steps

1. Deposit into a vault. Your first time for an asset should require an approval transaction for the SS router contract 0x6a1e73f12018D8e5f966ce794aa2921941feB17E.
2. After approval, your deposit into the yearn vault should have gone through the above address and you should receive yearn vault tokens to your address. See this transaction for reference: https://etherscan.io/tx/0x5d251a6647f3930be507ffa482600c590fe0a1ed0345fbeca7cf725e21ab8b9c
3. You should be able to withdraw your yearn tokens and get back the underlying token. The withdraw will not go through the router contract but will be a direct interaction with the vault.

### CODE REVIEW

File: src/features/earn/providers/yearn/api/api.ts [hunk 1]
```diff
@@ -107,31 +109,53 @@ export class YearnVaultApi {
...
+  }
+
+  async getVaultId({
+    tokenContractAddress,
+    vaultAddress
+  }: {
+    tokenContractAddress: string
+    vaultAddress: string
+  }): Promise<number> {
+    const routerAddress = new this.web3.eth.Contract(ssRouterAbi, ssRouterContractAddress)
```
Comment (0xean) [inline src/features/earn/providers/yearn/api/api.ts:123]: nit: would rename variable here to something more clear.

File: src/features/earn/providers/yearn/api/api.ts [hunk 2]
```diff
@@ -163,19 +187,23 @@ export class YearnVaultApi {
...
   }
 
   async depositEstimatedGas(input: TxEstimatedGasInput): Promise<BigNumber> {
-    const { amountDesired, userAddress, vaultAddress } = input
-    const vaultContract = new this.web3.eth.Contract(yv2VaultAbi, vaultAddress)
-    const estimatedGas = await vaultContract.methods
-      .deposit(amountDesired.toString(), userAddress)
+    const { amountDesired, userAddress, tokenContractAddress, vaultAddress } = input
+
+    const routerContract = new this.web3.eth.Contract(ssRouterAbi, ssRouterContractAddress)
```
Comment (0xean) [inline src/features/earn/providers/yearn/api/api.ts:198]: Is there any reason not to instantiate the router contract in the constructor and store it vs creating a `new` instance for each function call?

**Review by 0xean — APPROVED** [review]

File: src/features/earn/providers/yearn/api/api.ts (hunk 1 above)
Comment (theoboldfrazier) [reply src/features/earn/providers/yearn/api/api.ts]: Yeah, that was misnamed, it's the router contract, not the address. I'm brought the contract into the constructor and am using that now.

File: src/features/earn/providers/yearn/api/api.ts (hunk 2 above)
Comment (theoboldfrazier) [reply src/features/earn/providers/yearn/api/api.ts]: done.

File: src/features/earn/providers/yearn/api/api.ts [hunk 3]
```diff
@@ -107,31 +111,52 @@ export class YearnVaultApi {
...
+    vaultAddress
+  }: {
+    tokenContractAddress: string
+    vaultAddress: string
+  }): Promise<number> {
+    const numVaults = await this.ssRouterContract.methods
+      .numVaults(this.checksumAddress(tokenContractAddress))
+      .call()
+    let id: number = 0
+    for (let i = 0; i <= numVaults && !id; i++) {
```
Comment (cjthompson) [inline src/features/earn/providers/yearn/api/api.ts:129]: `0` is falsy. Do a null/undefined check

**Review by cjthompson — APPROVED** [review]

File: src/features/earn/providers/yearn/api/api.ts (hunk 3 above)
Comment (theoboldfrazier) [reply src/features/earn/providers/yearn/api/api.ts]: Done

**Pull request merged**